### DISCUSSION TITLE
How to save history.state after refresh?

### DISCUSSION THREAD

**Discussion by ilienkors** [vercel/next.js#38871] [Help]
Hello, I need to get history.state that I set before page reload but nextjs rewrites it. So is there any solution for such situation?

**Comment by Babbili** [comment]
@ilienkors you can use `localStorage`, you can set the value of `state` anywhere in the app in `localStorage`
```jsx
    window.localStorage.setItem('STATE', JSON.stringify(state))
```
then you can get it from `localStorage` on page load with `useEffect` hook
```jsx
    const [state, setState] = useState([])

    useEffect(() => {
        let storage = window.localStorage.getItem('STATE')
        if(storage != null) {
            let storageParsed = JSON.parse(storage)
            setState(storageParsed)
        } else {
            // console.log('storage', storage)
        }
    }, [])
```

**Reply by ilienkors** [reply]
@Babbili thank you for your answer but unfortunately that is not my case.

I need to work with history state to control custom router with back and forward browser buttons. On every "step" in my app I pushState with step number to history, then when user clicks "back button" I listen it with onpopstate event and change page to previous (or forward). To understand which button is clicked I need to compare what is in history.state and what is current step number in my react state. It can only be done with values in history. After page reload I have to history.go(-n) steps to clear history of clicking so user can use this feature after refresh. 

Localstorage has same value on different tabs what is not my option. I think I can use SessionStorage after reload and history.state while navigating but don't understand why nextjs developers removed possibility to work with history.state after page refresh. It is very useful thing for uncommon tasks. What is the reason to not just add their properies to mine like this?

`history.replaceState({...history.state, ...nextjsProperties})`

**Reply by Babbili** [reply]
@ilienkors i guess [Shallow Routing](https://nextjs.org/docs/routing/shallow-routing) in `next/router` would be the answer.
It will allow you to change the URL ( `/?step=1` , `/?step=2`, ... ) without running data fetching methods again and without losing state
```jsx
import { useRouter } from 'next/router'

// Current URL is '/'
function Page() {
  const router = useRouter()

  function nextBtnClick() {
    router.push('/?step=2', { shallow: true })
  }
}
```
You can also watch for URL changes via [componentDidUpdate](https://reactjs.org/docs/react-component.html#componentdidupdate) as shown below:
```jsx
componentDidUpdate(prevProps) {
  const { pathname, query } = this.props.router
  
  if (query.step !== prevProps.router.query.step) {
    // fetch data based on the new query
  }
}
```
see her full [Shallow Routing Docs](https://nextjs.org/docs/routing/shallow-routing) and [example](https://github.com/vercel/next.js/tree/canary/examples/with-shallow-routing)

**Reply by ilienkors** [reply]
@Babbili unfortunately I can't use query params because it wil cause problems when user go to the link with step param and I will do history.go(-stepsCount) while his history is empty.

So what I decided is to use history.state while using my app and save number of passed steps to session storage which I can use after refresh.

But question still be opened - why nexjs entirely rewrites history.state?

**Reply by acderusha** [reply]
I am having the same issue. I see that on page load, or even forward and back navigation, next is rewriting the history state when I would like to access custom data I pass in there.

Repro Steps:
1. Go to https://jobs.netflix.com/ (just because I know next is used there)
2. In the developer tools console type: `window.history.replaceState({...window.history.state, custom: 'myCustomHistoryValue'}, '', window.location.href)` to populate a custom state value
3. Refresh the page or go to a new page and hit the back button
4. Observe that the `custom: 'myCustomHistoryValue` value is no longer in window.history.state

I've been able to repro this on Chrome, Firefox, Safari and other browses. Is there a way to retain my custom history state here, and why is next rewriting the state? I would really like to avoid populating local and /or session storage here. Thanks!

**Reply by giorgiabosello** [reply]
I need to write in history state to, did you find a solution?

**Reply by Dev1OptimusTecno** [reply]
looks like this version introduced experimental suport for changing the history state 
https://github.com/vercel/next.js/releases/tag/v14.0.3
https://github.com/vercel/next.js/pull/58335
